File: java/src/test/java/com/showmeyourcode/playground/java/code/exercise/misc/TuringMachineExerciseTest.java
package com.showmeyourcode.playground.java.code.exercise.misc;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;

class TuringMachineExerciseTest {

    @Test
    void shouldPrintTapeWithoutErrors() {
        assertDoesNotThrow(() -> new TuringMachineExercise("1100").printTape());
    }

    @ParameterizedTest
    @CsvSource({
            "1,0",
            "10,01",
            "100,011",
            "1100,1011",
            "1110,1101"
    })
    void shouldFindPredecessor(String input, String predecessor) {
        var machine = new TuringMachineExercise(input);

        machine.run();

        assertEquals(predecessor, machine.getTape());
    }
}

File: kotlin/src/test/kotlin/com/showmeyourcode/playground/kotlin/code/exercise/misc/TuringMachineExerciseTest.kt
package com.showmeyourcode.playground.kotlin.code.exercise.misc

import io.kotest.assertions.throwables.shouldNotThrowAny
import io.kotest.core.spec.style.AnnotationSpec
import io.kotest.data.blocking.forAll
import io.kotest.data.row
import io.kotest.matchers.shouldBe

class TuringMachineExerciseTest : AnnotationSpec() {
    @Test
    fun `should print the tape without errors`() {
        shouldNotThrowAny {
            TuringMachineExercise("100").printTape()
        }
    }

    @Test
    fun `should find the predecessor`() {
        forAll(
            row("1", "0"),
            row("10", "01"),
            row("100", "011"),
            row("1100", "1011"),
            row("1110", "1101")
        ) { input, predecessor ->
            val machine = TuringMachineExercise(input)

            machine.run()

            machine.getTape() shouldBe predecessor
        }
    }
}
